### PR TITLE
Update dependencies

### PR DESCRIPTION
**Pull request by snlacks**
Switch from mout to lodash, updated the relevant code.
Updated eslint and tape, verified.

https://travis-ci.org/snlacks/credential

### CODE REVIEW

**Comment by ericelliott** [comment]
Please add `npm-debug.log` to `.gitignore` =)

File: package.json [hunk 1]
```diff
@@ -76,8 +80,6 @@
     "url": "https://github.com/dilvie/credential/issues"
   },
   "pre-commit": [
-    "lint",
-    "validate",
-    "test"
+    "validate"
```
Comment (ericelliott) [inline package.json]: Change this to lint. We want `pre-commit` to be really fast. Travis-CI will catch the rest. =)

**Comment by ericelliott** [comment]
This is starting to look really good! :cake:

File: package.json (hunk 1 above)
Comment (snlacks) [inline package.json]: While I'm in there, should I change the url for bugs just above it?

**Comment by ericelliott** [comment]
Uh-oh.

```
# constantEquals exposes no timings
not ok 27 ttest diff set is same as equal set
  ---
    operator: ok
    expected: true
    actual:   false
    at: Test.<anonymous> (/home/travis/build/ericelliott/credential/test/credential-test.js:237:5)
  ...
not ok 28 ttest diff set is same as inequal set
  ---
    operator: ok
    expected: true
    actual:   false
    at: Test.<anonymous> (/home/travis/build/ericelliott/credential/test/credential-test.js:238:5)
  ...
ok 29 ttest inequal set is same as equal set
```

That may be a security vulnerability, or it may just be Travis' build machine performing strangely?

Re-running...

**Comment by snlacks** [comment]
image

Not sure why the travis-ci build is failing now, it's failing on node .10 on travis-ci test https://travis-ci.org/snlacks/credential (first test #27 only, second test #27, 28), it's failing on iojs on yours (#26, 27).  Locally it tests fine for me on node and iojs. These are the constEquals test, any ideas?

**Comment by ericelliott** [comment]
The constantEquals tests use statistical timings to determine wether or not an attacker could use timing differences in password attempts in order to guess a password in a similar way to guessing a word playing hangman...

But I'm not even convinced that this hash algorithm is vulnerable to timing attacks in the first place. I'm just gun shy about removing it because I am far from a security expert, and I'm not sure whether or not the hash algorithm could be exploited in a non-obvious way with timing attacks.

**Comment by ericelliott** [comment]
@brianmhunt - Could you provide any help on this?

**Comment by brianmhunt** [comment]
There are a few options that would make the test more lenient, including at

[constantEquals exposes no timings](https://github.com/ericelliott/credential/blob/master/test/credential-test.js#L233) increasing `mu`, increasing `alpha`, or increasing the [iterations](https://github.com/ericelliott/credential/blob/master/test/credential-test.js#L219).

The first two (`mu` and `alpha`) will reduce the sensitivity (and strength) of the test. The latter (iterations) will increase the strength of the test but the test will take longer. I would not change `mu` as it is based on the BlackHat number (number of ms where something is detectible). You can change all three – the differences in whether this regularly passes or fails will depend on the processor, computer load, compiler, etc.

There is always a statistical chance that the test will fail, and the above settings were probably geared towards making the results on my machine pass. Lessening the sensitivity and/or increasing the iterations should make it pass more universally. Shooting from the hip, I would consider reducing `alpha` to `0.01`, and increasing the iterations to 3000, and see how far that gets it on most runs.

I am with @ericelliott, wondering whether the timing attack is a big concern – that said, better safe than sorry for this sort of thing. That said, until WebCrypto gets out, all the Javascript functions are at the mercy of the compiler optimizations, which are generally at odds with the objective of a constant-equals function.

**Comment by snlacks** [comment]
Could it be the test?

**Comment by brianmhunt** [comment]
@snlacks Sporadic failures likely indicate a problem with probability, and not one inherent to the test. (i.e. the test design is sound, but the statistical significance of the results must be adjusted to be practical)

**Comment by tjconcept** [comment]
How about using modular lodash: https://www.npmjs.com/package/lodash.assign? Or even https://www.npmjs.com/package/object-assign

**Comment by ericelliott** [comment]
Would it be crazy to abandon the idea of making the compare constant time, make this async, and just always take 500 ms to return?

**Comment by ericelliott** [comment]
don't worry about lodash for now. Different ticket. :)

**Comment by snlacks** [comment]
Lodash and the rest are already in this commit. I admittedly know only what I learned from this conversation and from Google over the last two days. I was thinking the changes are what started causing slight timing differences. Maybe some optimization or something.  There's several timing packages out there,  I can try them out this evening and see if any of them can pass regularly.

**Comment by tjconcept** [comment]
@ericelliott to be honest, I don't know, but 500ms is far too long (you would have to add that to time the hash requires).

**Comment by ericelliott** [comment]
@tjconcept want to take on the async constant delay? Feel free to pick a more appropriate fixed time, and make sure the statistical timing tests pass. :)

**Comment by tjconcept** [comment]
Actually I have my doubts that we are even vulnerable to timing attacks, and it is such a large (relative to the rest of the source) and fragile "moving part" that I would rather remove it from the source.
Having said so, I do not want to make that call myself.
In my private rewrite (which I use extensively in production) I have replaced it with the `scmp` package so far. Because I have a couple of high profile clients I decided to buy a security review, and my hopes are that his findings would apply here as well. You can track it here: https://github.com/srcagency/credential/issues/1

**Comment by ericelliott** [comment]
Great. :+1:

**Comment by ericelliott** [comment]
Sorry, this has gone stale. Can you update it?

**Pull request closed without merging**